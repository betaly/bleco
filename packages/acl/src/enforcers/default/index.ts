export * from './default-enforcer';
export * from './default-enforcer.provider';
export * from './policies';
