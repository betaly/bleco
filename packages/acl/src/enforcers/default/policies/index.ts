export * from './builder';
export * from './compiler';
export * from './linker';
export * from './types';
