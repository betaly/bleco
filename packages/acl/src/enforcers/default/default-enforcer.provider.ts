import {RepositoryFactory} from '@bleco/repository-factory';
import {BindingScope, inject, injectable, Provider} from '@loopback/context';
import {EnforcerStrategy} from '../../../dist';
import {Enforcer} from '../../enforcer';
import {AclBindings} from '../../keys';
import {PolicyRegistry} from '../../policies';
import {DefaultEnforcerOptions, DefaultEnhancer} from './index';

@injectable({scope: BindingScope.SINGLETON})
export class DefaultEnforcerProvider implements Provider<Enforcer> {
  constructor(
    @inject(AclBindings.POLICY_REGISTRY)
    readonly policyRegistry: PolicyRegistry,
    @inject(AclBindings.REPOSITORY_FACTORY)
    readonly repositoryFactory: RepositoryFactory,
    @inject(AclBindings.DEFAULT_ENFORCER_OPTIONS, {optional: true})
    readonly options: DefaultEnforcerOptions = {},
  ) {}

  value(): EnforcerStrategy {
    return new DefaultEnhancer(this.policyRegistry, this.repositoryFactory, this.options);
  }
}
