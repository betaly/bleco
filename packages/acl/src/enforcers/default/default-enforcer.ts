import {RepositoryFactory} from '@bleco/repository-factory';
import {BindingScope, injectable} from '@loopback/context';
import {AnyObject, Entity, EntityCrudRepository, FilterExcludingWhere, Options, Where} from '@loopback/repository';
import debugFactory from 'debug';
import {get} from 'lodash';
import {toArray} from 'tily/array/toArray';
import isEmpty from 'tily/is/empty';
import {Constructor} from 'tily/typings/types';
import {inspect} from 'util';
import {AuthorizedFilter, Enforcer} from '../../enforcer';
import {ForbiddenError, NotFoundError} from '../../errors';
import {toPrincipalPolymorphic, toResourcePolymorphic} from '../../helpers';
import {AclRoleMapping, AclRoleMappingRelations, AclRolePermission} from '../../models';
import {AclModelRelationKeys, PolicyRegistry, ResolvedPolicy, resolveModelName} from '../../policies';
import {buildRelationRolesIncludes, LocalRoleKey, PolicyBuilder} from './policies';
import ResourceRoles = AclModelRelationKeys.ResourceRoles;
import ResourceRoleMappings = AclModelRelationKeys.ResourceRoleMappings;

const debug = debugFactory('bleco:acl:default-enforcer');

export const DEFAULT_READ_ACTION = 'read';

export interface DefaultEnforcerOptions {
  readAction?: string;
}

@injectable({scope: BindingScope.SINGLETON})
export class DefaultEnhancer implements Enforcer {
  name = 'default';

  protected policyBuilder: PolicyBuilder;

  private readonly readAction: string;

  constructor(
    policyRegistry: PolicyRegistry,
    private repositoryFactory: RepositoryFactory,
    options: DefaultEnforcerOptions = {},
  ) {
    this.policyBuilder = new PolicyBuilder(policyRegistry.policies);
    this.readAction = options.readAction ?? DEFAULT_READ_ACTION;
  }

  async isAllowed(principal: Entity, action: string, resource: Entity): Promise<boolean> {
    const {_, ...roles} = this.policyBuilder.get(resource).actionRoles[action];

    let where: Where;
    let repo: EntityCrudRepository<Entity, unknown>;
    if (isEmpty(roles)) {
      repo = await this.repositoryFactory.getRepository(AclRoleMapping.name);
      where = buildAllowedWhereForRoleMappings(principal, resource, action, _);
    } else {
      repo = await this.repositoryFactory.getRepository(resource.constructor.name);
      where = buildAllowedWhereForResource(principal, this.policyBuilder.get(resource), action);
    }

    debug(`isAllowed check in ${repo.entityClass.name}: %O`, where);
    const [found] = await repo.find({where, limit: 1});
    return !!found;
  }

  async authorize(principal: Entity, action: string, resource: Entity, options?: {checkRead?: boolean}): Promise<void> {
    if (await this.isAllowed(principal, action, resource)) {
      return;
    }

    let isNotFound = false;
    if (options?.checkRead) {
      if (action === this.readAction) {
        isNotFound = true;
      } else {
        const canRead = await this.isAllowed(principal, this.readAction, resource);
        if (!canRead) {
          isNotFound = true;
        }
      }
    }
    const E = isNotFound ? NotFoundError : ForbiddenError;
    throw new E();
  }

  async authorizedActions(
    principal: Entity,
    resource: Entity,
    options?: {allowWildcard?: boolean},
  ): Promise<Set<string | '*'>> {
    const resourceCls = resource.constructor as typeof Entity;
    const resourceRepo = await this.repositoryFactory.getRepository(resourceCls.name);
    const roleMappingRepo = await this.repositoryFactory.getRepository<AclRoleMapping, AclRoleMappingRelations>(
      AclRoleMapping.name,
    );

    const policy = this.policyBuilder.get(resource);

    const inclusion = buildRelationRolesIncludes(resource, policy.roles);
    debug(`authorizedActions - resolve resource relations with:`, inspect(inclusion, {depth: null, colors: true}));

    // resolve resource relations
    const resourceWithRelations = await resourceRepo.findById(resource.getId(), {
      include: inclusion,
    });

    const {_, ...relRoles} = policy.roles;
    const or: AnyObject = [{'role.id': {neq: null}}];
    if (_) {
      // `or` custom role conditions
      or.push({
        ...toResourcePolymorphic(resource),
        roleId: {inq: _},
      });
    }
    if (relRoles) {
      // `or` relation roles conditions
      for (const rel of Object.keys(relRoles)) {
        const relResource = get(resourceWithRelations, rel);
        if (relResource) {
          or.push({
            ...toResourcePolymorphic(relResource),
            roleId: {inq: relRoles[rel]},
          });
        }
      }
    }
    const where: Where = {
      ...toPrincipalPolymorphic(principal),
      or,
    };
    debug(`authorizedActions find role mappings in condition:`, inspect(where, {depth: null, colors: true}));

    const mappings = await roleMappingRepo.find({
      where,
      // include .role.permissions
      include: [
        {
          relation: 'role',
          scope: {
            include: ['permissions'],
          },
        },
      ],
    });
    debug('authorizedActions found role mappings: %O', mappings);

    const actions = new Set<string | '*'>();

    // resolve actions from builtin roles
    for (const key in policy.roleActions) {
      let [rel, role] = key.split(':');
      if (!role) {
        role = rel;
        rel = LocalRoleKey;
      }
      const r = rel === LocalRoleKey ? resource : get(resourceWithRelations, rel);
      const roleMapping = mappings.find(
        m => m.roleId === role && m.resourceId === r?.getId() && m.resourceType === resolveModelName(r?.constructor),
      );
      if (roleMapping) {
        policy.roleActions[key].forEach(a => actions.add(a));
      }
    }

    // resolve actions from custom roles
    for (const mapping of mappings) {
      const role = mapping.role;
      if (role?.permissions) {
        for (const action of role.permissions.map((p: AclRolePermission) => p.action)) {
          actions.add(action);
        }
      }
    }

    return actions;
  }

  async authorizedQuery(
    principal: Entity,
    action: string,
    resourceCls: Constructor<Entity>,
  ): Promise<AuthorizedFilter> {
    const where = buildAllowedWhereForResource(principal, this.policyBuilder.get(resourceCls), action);
    return {model: resourceCls.name, where};
  }

  async authorizedResources<T extends Entity>(
    principal: Entity,
    action: string,
    resourceCls: Constructor<T>,
    filter?: FilterExcludingWhere<T>,
    options?: Options,
  ): Promise<T[]> {
    const repo = await this.repositoryFactory.getRepository<T>(resourceCls.name);
    const where = buildAllowedWhereForResource(principal, this.policyBuilder.get(resourceCls), action);
    return repo.find({where, ...filter}, options);
  }

  authorizeField(principal: Entity, action: string, resource: Entity, field: string): Promise<void> {
    throw new Error('Method not implemented.');
  }

  authorizedFields(
    principal: Entity,
    action: string,
    resource: Entity,
    options?: {allowWildcard?: boolean},
  ): Promise<Set<string | '*'>> {
    throw new Error('Method not implemented.');
  }
}

function buildAllowedWhereForRoleMappings(
  principal: Entity,
  resource: Entity,
  action: string | string[],
  roles: string[],
) {
  const actions = toArray(action);
  const actionValue = actions.length === 1 ? actions[0] : {inq: actions};
  const or: AnyObject[] = [{'role.permissions.action': actionValue}];
  if (!isEmpty(roles)) {
    or.push({roleId: {inq: roles}});
  }
  return {
    ...toPrincipalPolymorphic(principal),
    ...toResourcePolymorphic(resource),
    or,
  } as Where;
}

function buildAllowedWhereForResource(principal: Entity, resourcePolicy: ResolvedPolicy, action: string) {
  const {principalType, principalId} = toPrincipalPolymorphic(principal);
  const {_, ...roles} = resourcePolicy.actionRoles[action];
  const or: AnyObject[] = [
    {
      [`${ResourceRoles}.principals.principalType`]: principalType,
      [`${ResourceRoles}.principals.principalId`]: principalId,
      [`${ResourceRoles}.permissions.action`]: action,
    },
  ];
  if (!isEmpty(_)) {
    or.push({
      [`${ResourceRoleMappings}.principalType`]: principalType,
      [`${ResourceRoleMappings}.principalId`]: principalId,
      [`${ResourceRoleMappings}.roleId`]: {inq: _},
    });
  }
  if (!isEmpty(roles)) {
    for (const rel of Object.keys(roles)) {
      or.push({
        [`${rel}.${ResourceRoleMappings}.principalType`]: principalType,
        [`${rel}.${ResourceRoleMappings}.principalId`]: principalId,
        [`${rel}.${ResourceRoleMappings}.roleId`]: {inq: roles[rel]},
      });
    }
  }
  return {or} as Where;
}
