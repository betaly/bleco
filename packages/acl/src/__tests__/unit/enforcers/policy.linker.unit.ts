import {compile, link} from '../../../enforcers/default';
import {IssuePolicy, OrgPolicy, RepoPolicy} from '../../../test';

describe('Policy Linker', function () {
  it('should link', async () => {
    const compiled = compile([OrgPolicy, RepoPolicy, IssuePolicy]);
    const resolved = link(compiled);
    expect(resolved).toMatchSnapshot();
  });
});
