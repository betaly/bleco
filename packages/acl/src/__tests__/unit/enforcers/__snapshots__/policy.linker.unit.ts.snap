// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Policy Linker should link 1`] = `
Array [
  Object {
    "actionRoles": Object {
      "create_orgs": Object {
        "_": Array [
          "member",
          "admin",
        ],
      },
      "manage": Object {
        "_": Array [
          "admin",
        ],
      },
    },
    "actions": Array [
      "manage",
      "create_orgs",
    ],
    "definition": Object {
      "actions": Array [
        "manage",
        "create_orgs",
      ],
      "model": [Function],
      "name": "Global",
      "roleActions": Object {
        "admin": Array [
          "manage",
        ],
        "member": Array [
          "create_orgs",
        ],
      },
      "roleDerivations": Object {
        "member": Array [
          "admin",
        ],
      },
      "roles": Array [
        "admin",
        "member",
      ],
      "type": "resource",
    },
    "model": [Function],
    "roleActions": Object {
      "_:admin": Array [
        "manage",
        "create_orgs",
      ],
      "_:member": Array [
        "create_orgs",
      ],
    },
    "roleChildren": Object {
      "admin": Object {
        "_": Array [],
      },
      "member": Object {
        "_": Array [
          "admin",
        ],
      },
    },
    "roleParents": Object {
      "admin": Array [
        "member",
      ],
      "member": Array [],
    },
    "roles": Object {
      "_": Array [
        "admin",
        "member",
      ],
    },
  },
  Object {
    "actionRoles": Object {
      "create_repos": Object {
        "_": Array [
          "owner",
        ],
        "global": Array [
          "admin",
        ],
      },
      "create_role_assignments": Object {
        "_": Array [
          "owner",
        ],
        "global": Array [
          "admin",
        ],
      },
      "delete_role_assignments": Object {
        "_": Array [
          "owner",
        ],
        "global": Array [
          "admin",
        ],
      },
      "list_repos": Object {
        "_": Array [
          "member",
          "owner",
        ],
        "global": Array [
          "admin",
        ],
      },
      "list_role_assignments": Object {
        "_": Array [
          "member",
          "owner",
        ],
        "global": Array [
          "admin",
        ],
      },
      "read": Object {
        "_": Array [
          "member",
          "owner",
        ],
        "global": Array [
          "admin",
        ],
      },
      "update_role_assignments": Object {
        "_": Array [
          "owner",
        ],
        "global": Array [
          "admin",
        ],
      },
    },
    "actions": Array [
      "read",
      "create_repos",
      "list_repos",
      "create_role_assignments",
      "list_role_assignments",
      "update_role_assignments",
      "delete_role_assignments",
    ],
    "definition": Object {
      "actions": Array [
        "read",
        "create_repos",
        "list_repos",
        "create_role_assignments",
        "list_role_assignments",
        "update_role_assignments",
        "delete_role_assignments",
      ],
      "model": [Function],
      "name": "Org",
      "relations": Array [
        "global",
      ],
      "roleActions": Object {
        "member": Array [
          "read",
          "list_repos",
          "list_role_assignments",
        ],
        "owner": Array [
          "create_repos",
          "create_role_assignments",
          "update_role_assignments",
          "delete_role_assignments",
        ],
      },
      "roleDerivations": Object {
        "member": Array [
          "owner",
        ],
        "owner": Array [
          "global.admin",
        ],
      },
      "roles": Array [
        "owner",
        "member",
      ],
      "type": "resource",
    },
    "model": [Function],
    "roleActions": Object {
      "_:member": Array [
        "read",
        "list_repos",
        "list_role_assignments",
      ],
      "_:owner": Array [
        "read",
        "create_repos",
        "list_repos",
        "create_role_assignments",
        "list_role_assignments",
        "update_role_assignments",
        "delete_role_assignments",
      ],
      "global:admin": Array [
        "read",
        "create_repos",
        "list_repos",
        "create_role_assignments",
        "list_role_assignments",
        "update_role_assignments",
        "delete_role_assignments",
      ],
    },
    "roleChildren": Object {
      "member": Object {
        "_": Array [
          "owner",
        ],
        "global": Array [
          "admin",
        ],
      },
      "owner": Object {
        "_": Array [],
        "global": Array [
          "admin",
        ],
      },
    },
    "roleParents": Object {
      "global.admin": Array [
        "owner",
        "member",
      ],
      "member": Array [],
      "owner": Array [
        "member",
      ],
    },
    "roles": Object {
      "_": Array [
        "member",
        "owner",
      ],
      "global": Array [
        "admin",
      ],
    },
  },
  Object {
    "actionRoles": Object {
      "create_issues": Object {
        "_": Array [
          "reader",
          "admin",
          "maintainer",
        ],
        "org": Array [
          "owner",
          "member",
        ],
        "org.global": Array [
          "admin",
        ],
      },
      "create_role_assignments": Object {
        "_": Array [
          "admin",
        ],
        "org": Array [
          "owner",
        ],
        "org.global": Array [
          "admin",
        ],
      },
      "delete_role_assignments": Object {
        "_": Array [
          "admin",
        ],
        "org": Array [
          "owner",
        ],
        "org.global": Array [
          "admin",
        ],
      },
      "list_issues": Object {
        "_": Array [
          "reader",
          "admin",
          "maintainer",
        ],
        "org": Array [
          "owner",
          "member",
        ],
        "org.global": Array [
          "admin",
        ],
      },
      "list_role_assignments": Object {
        "_": Array [
          "admin",
        ],
        "org": Array [
          "owner",
        ],
        "org.global": Array [
          "admin",
        ],
      },
      "read": Object {
        "_": Array [
          "reader",
          "admin",
          "maintainer",
        ],
        "org": Array [
          "owner",
          "member",
        ],
        "org.global": Array [
          "admin",
        ],
      },
      "update_role_assignments": Object {
        "_": Array [
          "admin",
        ],
        "org": Array [
          "owner",
        ],
        "org.global": Array [
          "admin",
        ],
      },
    },
    "actions": Array [
      "read",
      "create_issues",
      "list_issues",
      "create_role_assignments",
      "list_role_assignments",
      "update_role_assignments",
      "delete_role_assignments",
    ],
    "definition": Object {
      "actions": Array [
        "read",
        "create_issues",
        "list_issues",
        "create_role_assignments",
        "list_role_assignments",
        "update_role_assignments",
        "delete_role_assignments",
      ],
      "model": [Function],
      "name": "Repo",
      "relations": Array [
        "org",
      ],
      "roleActions": Object {
        "admin": Array [
          "create_role_assignments",
          "list_role_assignments",
          "update_role_assignments",
          "delete_role_assignments",
        ],
        "reader": Array [
          "read",
          "list_issues",
          "create_issues",
        ],
      },
      "roleDerivations": Object {
        "admin": Array [
          "org.owner",
        ],
        "maintainer": Array [
          "admin",
        ],
        "reader": Array [
          "org.member",
          "maintainer",
        ],
      },
      "roles": Array [
        "admin",
        "maintainer",
        "reader",
      ],
      "type": "resource",
    },
    "model": [Function],
    "roleActions": Object {
      "_:admin": Array [
        "read",
        "create_issues",
        "list_issues",
        "create_role_assignments",
        "list_role_assignments",
        "update_role_assignments",
        "delete_role_assignments",
      ],
      "_:maintainer": Array [
        "read",
        "create_issues",
        "list_issues",
      ],
      "_:reader": Array [
        "read",
        "create_issues",
        "list_issues",
      ],
      "org.global:admin": Array [
        "read",
        "create_issues",
        "list_issues",
        "create_role_assignments",
        "list_role_assignments",
        "update_role_assignments",
        "delete_role_assignments",
      ],
      "org:member": Array [
        "read",
        "create_issues",
        "list_issues",
      ],
      "org:owner": Array [
        "read",
        "create_issues",
        "list_issues",
        "create_role_assignments",
        "list_role_assignments",
        "update_role_assignments",
        "delete_role_assignments",
      ],
    },
    "roleChildren": Object {
      "admin": Object {
        "_": Array [],
        "org": Array [
          "owner",
        ],
      },
      "maintainer": Object {
        "_": Array [
          "admin",
        ],
        "org": Array [
          "owner",
        ],
      },
      "reader": Object {
        "_": Array [
          "admin",
          "maintainer",
        ],
        "org": Array [
          "owner",
          "member",
        ],
      },
    },
    "roleParents": Object {
      "admin": Array [
        "maintainer",
        "reader",
      ],
      "maintainer": Array [
        "reader",
      ],
      "org.member": Array [
        "reader",
      ],
      "org.owner": Array [
        "admin",
        "maintainer",
        "reader",
      ],
      "reader": Array [],
    },
    "roles": Object {
      "_": Array [
        "reader",
        "admin",
        "maintainer",
      ],
      "org": Array [
        "owner",
        "member",
      ],
      "org.global": Array [
        "admin",
      ],
    },
  },
  Object {
    "actionRoles": Object {
      "close": Object {
        "_": Array [
          "creator",
        ],
        "repo": Array [
          "maintainer",
          "admin",
        ],
        "repo.org": Array [
          "owner",
        ],
        "repo.org.global": Array [
          "admin",
        ],
      },
      "read": Object {
        "_": Array [],
        "repo": Array [
          "reader",
          "admin",
          "maintainer",
        ],
        "repo.org": Array [
          "owner",
          "member",
        ],
        "repo.org.global": Array [
          "admin",
        ],
      },
    },
    "actions": Array [
      "read",
      "close",
    ],
    "definition": Object {
      "actions": Array [
        "read",
        "close",
      ],
      "model": [Function],
      "name": "Issue",
      "relations": Array [
        "repo",
      ],
      "roleActions": Object {
        "creator": Array [
          "close",
        ],
        "repo.maintainer": Array [
          "close",
        ],
        "repo.reader": Array [
          "read",
        ],
      },
      "roles": Array [
        "creator",
      ],
      "type": "resource",
    },
    "model": [Function],
    "roleActions": Object {
      "_:creator": Array [
        "close",
      ],
      "repo.org.global:admin": Array [
        "read",
        "close",
      ],
      "repo.org:member": Array [
        "read",
      ],
      "repo.org:owner": Array [
        "read",
        "close",
      ],
      "repo:admin": Array [
        "read",
        "close",
      ],
      "repo:maintainer": Array [
        "read",
        "close",
      ],
      "repo:reader": Array [
        "read",
      ],
    },
    "roleChildren": Object {
      "creator": Object {
        "_": Array [],
      },
    },
    "roleParents": Object {
      "creator": Array [],
    },
    "roles": Object {
      "_": Array [
        "creator",
      ],
      "repo": Array [
        "reader",
        "admin",
        "maintainer",
      ],
      "repo.org": Array [
        "owner",
        "member",
      ],
      "repo.org.global": Array [
        "admin",
      ],
    },
  },
]
`;
